wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/convert/tool/kubernetes/KubernetesToolToWorkspaceApplier.java

```diff
@@ -86,24 +94,113 @@ public void apply(
+    estimateCommandsMachineName(workspaceConfig, k8sTool, toolObjects);
 
-    applyEntrypoints(k8sTool.getEntrypoints(), list);
+    applyEntrypoints(k8sTool.getEntrypoints(), toolObjects);
 
-    RecipeImpl recipe =
-        new RecipeImpl(k8sTool.getType(), YAML_CONTENT_TYPE, asYaml(k8sTool, list), null);
+    provisionEnvironment(workspaceConfig, k8sTool, toolObjects);
+  }
+
+  /**
+   * Provision default environment with the specified kubernetes/openshift tool.
+   *
+   * <p>If there is already a default environment with kubernetes/openshift then content will be
+   * updated with new list.
+   *
+   * @param workspaceConfig workspace where recipe should be provisioned
+   * @param k8sTool kubernetes/openshift tool that should be provisioned
+   * @param toolObjects parsed objects of the specified tool
+   * @throws DevfileRecipeFormatException if exception occurred during existing environment parsing
+   * @throws DevfileRecipeFormatException if exception occurred during kubernetes object
+   *     serialization
+   * @throws DevfileException if any other exception occurred
+   */
+  private void provisionEnvironment(
+      WorkspaceConfigImpl workspaceConfig, Tool k8sTool, List<HasMetadata> toolObjects)
+      throws DevfileException, DevfileRecipeFormatException {
+    String defaultEnv = workspaceConfig.getDefaultEnv();
+    EnvironmentImpl environment = workspaceConfig.getEnvironments().get(defaultEnv);
+    if (environment == null) {
+      RecipeImpl recipe =
+          new RecipeImpl(k8sTool.getType(), YAML_CONTENT_TYPE, asYaml(k8sTool, toolObjects), null);
+      String envName = k8sTool.getName();
+      workspaceConfig.getEnvironments().put(envName, new EnvironmentImpl(recipe, emptyMap()));
+      workspaceConfig.setDefaultEnv(envName);
+    } else {
+      RecipeImpl envRecipe = environment.getRecipe();
+
+      // check if it is needed to update recipe type since
+      // kubernetes tool is compatible with openshift but not vice versa
+      if (OPENSHIFT_TOOL_TYPE.equals(k8sTool.getType())
+          && KubernetesEnvironment.TYPE.equals(envRecipe.getType())) {
+        envRecipe.setType(OpenShiftEnvironment.TYPE);
+      }
+
+      // workspace already has k8s/OS recipe
+      // it is needed to merge existing recipe objects with tool's ones
+      List<HasMetadata> envObjects = unmarshalDefaultEnvObjects(workspaceConfig);
+      fillIn(envObjects, toolObjects);
+
+      envRecipe.setContent(asYaml(k8sTool, envObjects));
+    }
+  }
+
+  /**
+   * Fill in the specified target list with the specified objects.
+   *
+   * @param target list that should be filled in
+   * @param objects objects to fill in
+   * @throws DevfileFormatException if objects list contains item with no unique combination of kind
+   *     and name
+   */
+  private void fillIn(List<HasMetadata> target, List<HasMetadata> objects)
+      throws DevfileFormatException {
+    Set<String> uniqueNames = new HashSet<>();
```

Comment:
@metlos WDYT about using `Set<Pair<String, String>>` here ?
https://github.com/eclipse/che/blob/07263f1e30089689d71b057f747a44a29283e3c4/core/commons/che-core-commons-lang/src/main/java/org/eclipse/che/commons/lang/Pair.java#L15

Reply:
Well, you need to do a lookup. If it is a `Map<String, Set<String>>`, you do two lookups in smaller maps. If you have `Set<Pair<String, String>>` you do 1 lookup in a larger set + instantiation of a throwaway pair. Not really sure what's better, will let you judge.

Reply:
We do like 16 requests to DB on 1 workspace start, so I do not think that using `Map<String, Set<String>>` may optimize something ) 
I chose using Set<Pair> because the code looks much simpler with it.